### PR TITLE
Fixed so beacon is updated properly when props/persistent state change

### PR DESCRIPTION
**Pull request by matthewgertner**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-beacon",
-  "version": "0.8.3",
+  "version": "0.8.6",
```
Comment (matthewgertner) [inline package.json]: I haven't been a bit loose about updating the version numbers, only doing it when I publish, which is why they aren't always consecutive in the pull requests.

File: src/Beacon.jsx [hunk 2]
```diff
@@ -14,16 +14,17 @@ const TOOLTIP_TOLERANCE = 50;
 
 function deepCloneNode(node, parent, fn) {
   const clone = fn(node, parent);
-  const childNodes = [...node.childNodes];
+  const childNodes = Array.prototype.slice.call(node.childNodes);
```
Comment (matthewgertner) [inline src/Beacon.jsx:18]: Not sure why these changes are duplicated here.

File: src/Beacon.jsx [hunk 3]
```diff
@@ -47,31 +48,12 @@ export class Beacon extends Component {
...
-        const transaction = this.state.database.transaction(['beacons']);
-        const objectStore = transaction.objectStore('beacons');
-        objectStore.get(hash).onsuccess = (event) => {
-          if (!event.target.result) {
-            this.setState({ hash });
-          }
-        };
-      };
-    }
+    this.checkHash();
```
Comment (matthewgertner) [inline src/Beacon.jsx:57]: Moved this to a separate function since we need to do it when we update as well.

File: src/Beacon.jsx [hunk 4]
```diff
@@ -87,7 +69,16 @@ export class Beacon extends Component {
     }
   }
 
+  componentWillReceiveProps(nextProps) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:73]: Need to update the `inactive` state whenever we receive new props.

File: src/Beacon.jsx [hunk 5]
```diff
@@ -176,8 +167,9 @@ export class Beacon extends Component {
   }
 
   renderBeacon(persistent) {
-    if (persistent && !this.state.hash) {
-      // We need to wait until the hash is set before we render.
+    if (this.state.inactive || (persistent && !this.state.hash)) {
+      // If the beacon is inactive then don't render it.
+      // Otherwise we need to wait until the hash is set before we render.
```
Comment (matthewgertner) [inline src/Beacon.jsx:183]: Comment updated

File: src/Beacon.jsx [hunk 6]
```diff
@@ -87,7 +69,16 @@ export class Beacon extends Component {
...
 
+  componentWillReceiveProps(nextProps) {
+    const inactive = (nextProps.active === false) || (this.context.beacon && (this.context.beacon.active === false));
+    if (inactive !== this.state.inactive) {
+      this.setState({ inactive });
+    }
+  }
+
   componentWillUpdate() {
+    this.checkHash();
```
Comment (matthewgertner) [inline src/Beacon.jsx:81]: So now we check this here as well.

File: src/Beacon.jsx [hunk 7]
```diff
@@ -242,6 +236,48 @@ export class Beacon extends Component {
     }
   }
 
+  loadHash(persistent) {
```
Comment (matthewgertner) [inline src/Beacon.jsx:250]: So every time we mount/update, we check whether the hash is in the database. If it isn't, we set `this.state.hash` so that the beacon will render. If it is, we clear it so the beacon doesn't render (anymore).

File: assets/style.styl [hunk 8]
```diff
@@ -263,6 +253,7 @@ for $name, $motion in $animBank
 // Tether
 
 .tether-element
+  position fixed !important // absolute creates a scrollable body overflow bug in Android
```
Comment (matthewgertner) [inline assets/style.styl:256]: Unrelated Android fix

File: src/Beacon.jsx [hunk 9]
```diff
@@ -106,6 +98,16 @@ export class Beacon extends Component {
...
+    const targetElement = this.getTargetElement();
+    const targetBounds = targetElement.getBoundingClientRect();
+    const boundsChanged =
+      this.state.targetBounds.top !== targetBounds.top ||
+      this.state.targetBounds.bottom !== targetBounds.bottom ||
+      this.state.targetBounds.left !== targetBounds.left ||
+      this.state.targetBounds.right !== targetBounds.right;
+    if (boundsChanged || (this.state.targetElement !== targetElement)) {
+      this.setState({ targetElement, targetBounds }); // eslint-disable-line react/no-did-update-set-state
+    }
```
Comment (matthewgertner) [inline src/Beacon.jsx:110]: Same deal here... turns out the bounds can change on update (e.g. when using ReactSlider).

File: src/Beacon.jsx [hunk 10]
```diff
@@ -242,6 +247,50 @@ export class Beacon extends Component {
     }
   }
 
+  loadHash(persistent) {
+    const { tooltipText } = this.props;
+    const text = typeof(tooltipText) === 'string' ? tooltipText : ReactDOMServer.renderToStaticMarkup(tooltipText);
```
Comment (matthewgertner) [inline src/Beacon.jsx:252]: Can't just use `JSON.stringify` here if a `ReactElement` is passed in.

Comment (tomkis) [inline src/Beacon.jsx:252]: interesting, yet I am afraid of breaking this in next versions of React. `renderToStaticMarkup` is server specific so it might potentially use some features which are not in the browser.

File: src/Beacon.jsx [hunk 11]
```diff
@@ -242,6 +247,50 @@ export class Beacon extends Component {
...
+        }
+      } else {
+        if (this.state.hash) {
+          this.setState({ hash: null });
+        }
+      }
+    };
+  }
+
+  checkHash() {
```
Comment (tomkis) [inline src/Beacon.jsx:270]: `checkHash` is called in `componentWillUpdate` yet you are calling `setState` inside, I know that now it's asynchronous so it shouldn't matter, but what if you changed the implementation later to something synchronous? The real point is, should it really be used in `componentWillUpdate` ?

**Comment by matthewgertner** [comment]
Merging this since I made a lot of changes (see next PR).
